drivers/dma/dma_silabs_siwx917.c

```diff
@@ -7,54 +7,61 @@
 #include <errno.h>
 #include <zephyr/irq.h>
 #include <zephyr/sys/util.h>
+#include <zephyr/sys/sys_io.h>
 #include <zephyr/device.h>
 #include <zephyr/drivers/dma.h>
 #include <zephyr/drivers/clock_control.h>
 #include <zephyr/logging/log.h>
+#include <zephyr/sys/mem_blocks.h>
 #include <zephyr/types.h>
 #include "rsi_rom_udma_wrapper.h"
+#include "rsi_rom_udma.h"
 #include "rsi_udma.h"
 #include "sl_status.h"
 
-#define DT_DRV_COMPAT          silabs_siwx917_dma
-#define DMA_MAX_TRANSFER_COUNT 1024
-#define DMA_CH_PRIORITY_HIGH   1
-#define DMA_CH_PRIORITY_LOW    0
-#define VALID_BURST_LENGTH     0
-#define UDMA_ADDR_INC_NONE     0X03
+#define DT_DRV_COMPAT                    silabs_siwx917_dma
+#define DMA_MAX_TRANSFER_COUNT           1024
+#define DMA_CH_PRIORITY_HIGH             1
+#define DMA_CH_PRIORITY_LOW              0
+#define VALID_BURST_LENGTH               0
+#define UDMA_ADDR_INC_NONE               0x03
+#define UDMA_MODE_PER_ALT_SCATTER_GATHER 0x07
 
 LOG_MODULE_REGISTER(si91x_dma, CONFIG_DMA_LOG_LEVEL);
 
 struct dma_siwx917_config {
-	UDMA0_Type *reg;                 /* UDMA register base address */
-	uint8_t channels;                /* UDMA channel count */
-	uint8_t irq_number;              /* IRQ number */
-	RSI_UDMA_DESC_T *sram_desc_addr; /* SRAM Address for UDMA Descriptor Storage */
-	const struct device *clock_dev;
-	clock_control_subsys_t clock_subsys;
-	void (*irq_configure)(void);     /* IRQ configure function */
+	UDMA0_Type *reg;                     /* UDMA register base address */
+	uint8_t channels;                    /* UDMA channel count */
+	uint8_t irq_number;                  /* IRQ number */
+	RSI_UDMA_DESC_T *sram_desc_addr;     /* SRAM Address for UDMA Descriptor Storage */
+	const struct device *clock_dev;      /* Pointer to the clock device */
+	clock_control_subsys_t clock_subsys; /* Subsystem for clock control */
+	void (*irq_configure)(void);         /* IRQ configure function */
 };
 
 struct dma_siwx917_data {
 	UDMA_Channel_Info *chan_info;
-	dma_callback_t dma_callback;         /* User callback */
-	void *cb_data;                       /* User callback data */
-	RSI_UDMA_DATACONTEXT_T dma_rom_buff; /* Buffer to store UDMA handle */
-					     /* related information */
+	dma_callback_t dma_callback;          /* User callback */
+	void *cb_data;                        /* User callback data */
+	struct sys_mem_blocks *dma_desc_pool; /* Pointer to the memory pool for DMA descriptor */
+	RSI_UDMA_DATACONTEXT_T dma_rom_buff;  /* Buffer to store UDMA handle
+					       * related information
+					       */
 };
 
-static inline int siwx917_dma_is_peripheral_request(uint32_t dir)
+static int siwx917_transfer_direction(uint32_t dir)
 {
 	if (dir == MEMORY_TO_MEMORY) {
 		return 0;
 	}
 	if (dir == MEMORY_TO_PERIPHERAL || dir == PERIPHERAL_TO_MEMORY) {
 		return 1;
 	}
+	/* Invalid transfer direction */
 	return -1;
```

Comment:
It's not a good idea to use `-1` for `int` return values. The reason is, that pretty much everywhere in Zephyr the convention is to use negative POSIX error values (in our libc versions `EPERM` is defined as `1`, so `-1` could be translated to `-EPERM`). If you really need a negative error value here it seems like `-EINVAL` would be the right choice.